java-checks/src/main/java/org/sonar/java/checks/CognitiveComplexityMethodCheck.java

```diff
@@ -0,0 +1,476 @@
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.TryStatementTree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+import org.sonar.plugins.java.api.tree.WhileStatementTree;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.sonar.plugins.java.api.tree.Tree.Kind.*;
+
+
+@Rule(key = "S3776")
+public class CognitiveComplexityMethodCheck  extends IssuableSubscriptionVisitor {
+
+  private static final int DEFAULT_MAX = 10;
+
+  @RuleProperty(
+          key = "Threshold",
+          description = "The maximum authorized complexity.",
+          defaultValue = "" + DEFAULT_MAX)
+  private int max = DEFAULT_MAX;
+
+
+  private List<Tree.Kind> kindsAffectedByNesting = ImmutableList.<Tree.Kind>builder()
+          .add(IF_STATEMENT)
+          .add(CONDITIONAL_EXPRESSION)  // ternary
+          .add(FOR_STATEMENT)
+          .add(FOR_EACH_STATEMENT)
+          .add(DO_STATEMENT)
+          .add(WHILE_STATEMENT)
+          .add(SWITCH_STATEMENT)
+          .build();
+
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(METHOD, CONSTRUCTOR);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+
+    MethodTree method = (MethodTree) tree;
+    if (method.block() == null || ((ClassTree)method.parent()).simpleName() == null) {
+      return;
+    }
+
+    List<JavaFileScannerContext.Location> flow = new ArrayList<>();
+    int total = countComplexity(method.block().body(), flow, 0);
+
+    if (total > max) {
+      reportIssue(
+              method.simpleName(),
+              "Refactor this method to reduce its Cognitive Complexity from " + total + " to the " + max + " allowed.",
+              flow,
+              total - max);
+    }
+
+  }
+
+  private int countComplexity(List<StatementTree> statements, List<JavaFileScannerContext.Location> flow, int nestingLevel) {
+    if (statements.isEmpty()) {
+      return 0;
+    }
+
+    int total = 0;
+
+    for (StatementTree st : statements) {
+
+      Tree tree = getSegmentOfInterest(st);
+
+      if (kindsAffectedByNesting.contains(tree.kind())) {
+        int hit = 1 + nestingLevel;
+
+        addSecondaryLocation(flow, tree, hit, nestingLevel);
+
+        total += hit;
+        total += countExtraConditions(tree, flow);
+        total += countIfElseChains(tree, flow, nestingLevel);
+
+      } else {
+        total += countBreakAndContinue(tree, flow);
+        total += countTryChains(tree, flow, nestingLevel);
+        total += countAnonymousClasses(tree, flow, nestingLevel);
+      }
+
+      if (tree.is(TRY_STATEMENT)) {
+        total += countComplexity(getChildren(tree), flow, nestingLevel);
+      } else {
+        total += countComplexity(getChildren(tree), flow, nestingLevel + 1);
+      }
+    }
+
+    return total;
+  }
+
+  private static void addSecondaryLocation(List<JavaFileScannerContext.Location> flow, Tree st, int hit,  int nestingLevel) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) The Cyclomatic Complexity of this method "addSecondaryLocation" is 14 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)